File: leetcode/658-find-k-closest-elements/main.go
package main

import (
	"fmt"
)

// binary search
func bsearch(nums []int, target int) (int, int, int) {
	min := 0
	max := len(nums) - 1
	var mean int
	for min <= max {
		mean = (min + max) / 2
		if target == nums[mean] {
			return min, max, mean
		} else if target > nums[mean] {
			min = mean + 1
		} else if target < nums[mean] {
			max = mean - 1
		}
	}
	fmt.Println(min, max)
	return min, max, -1
}

func findClosestElements(arr []int, k int, x int) []int {
	if k >= len(arr) {
		return arr
	}
	min, max, mean := bsearch(arr, x)
	var target int
	if mean != -1 {
		target = mean
	} else if max == -1 {
		target = min
	} else if max == len(arr)-1 {
		target = max
	} else {
		if x-arr[max] <= arr[max+1]-x {
			target = max
		} else {
			target = max + 1
		}
	}
	fmt.Println("target", target)

	var left int
	var right int

	left = target
	right = target

	for right-left+1 < k {
	}

	return []int{left, right}
}

func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	a := []int{1, 2, 3, 4, 5}
	b := findClosestElements(a, 3, 6)
	fmt.Println(b)
}

File: leetcode/658-find-k-closest-elements/main.py
"""
    1st: nearest binary search

    Time        O(logN + K)
    Space       O(K)
    308 ms, faster than 61.39%
"""
import heapq


class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        n = len(arr)
        centerIdx = self.bSearchNearest(arr, x)
        left, right = centerIdx, centerIdx
        k -= 1
        while k > 0:
            if left-1 >= 0 and right+1 < n:
                if abs(arr[left-1] - x) < abs(arr[right+1] - x):
                    left -= 1
                elif abs(arr[left-1] - x) > abs(arr[right+1] - x):
                    right += 1
                else:
                    left -= 1
            elif left-1 < 0:
                right += 1
            elif right+1 == n:
                left -= 1
            k -= 1
        return arr[left:right+1]

    def bSearchNearest(self, nums, target):
        n = len(nums)
        left = 0
        right = n-1
        while left <= right:
            mid = (left + right)//2
            if target < nums[mid]:
                right = mid - 1
            elif target > nums[mid]:
                left = mid + 1
            else:
                return mid
        # bounds checking
        if right < 0:
            return 0
        if left > n-1:
            return n-1
        # compare
        if abs(target - nums[right]) <= abs(target - nums[left]):
            return right
        return left


"""
    2nd: heap

    Time    O(NlogN + KlogK)
    Space   O(N)
    400 ms, faster than 20.44%
"""


class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        pq = []
        for num in arr:
            diff = abs(num - x)
            heapq.heappush(pq, (diff, num))
        res = []
        while len(res) < k:
            diff, num = heapq.heappop(pq)
            res.append(num)
        return sorted(res)
